retrofit/src/main/java/retrofit2/RequestBuildInterceptor.java

```diff
@@ -0,0 +1,25 @@
+package retrofit2;
+
+import java.lang.reflect.Method;
+
+import okhttp3.Request;
+
+/**
+ * With this interceptor, you can intercept the request build process and modify
+ * the request as you want it to be executed. i.e. adding headers based on annotations
+ */
+public interface RequestBuildInterceptor {
+
+  /**
+   * @param request the request object as it has been generated based on the method
+   * @return a (modified) request, as you want it to be called
+   */
+  Request build(Request request);
+
+  /**
+   * Creates {@link RequestBuildInterceptor} based on the method which has been called
+   */
+  interface Factory {
+    RequestBuildInterceptor get(Method method);
```

Comment:
instead of passing the whole method object I would be ok with passing the annotations only, but who knows what the user actually wants to read out of the method. So I decided to pass the method, to be more open for other options as well (i.e. return type dependent requests or similar things)